src/attributes.jl

```diff
@@ -1102,6 +1102,30 @@ struct ConstraintBasisStatus <: AbstractConstraintAttribute
 end
 ConstraintBasisStatus() = ConstraintBasisStatus(1)
 
+"""
+    CanonicalConstraintFunction()
+
+A constraint attribute for a canonical representation of the
+[`AbstractFunction`](@ref) object used to define the constraint.
+Getting this attribute is guaranteed to return a function that is equivalent but
+not necessarily identical to the function provided by the user.
+
+By default, `MOI.get(model, MOI.CanonicalConstraintFunction(), ci)` fallbacks to
+`MOI.Utilities.canonical(MOI.get(model, MOI.ConstraintFunction(), ci))`.
+However, if `model` knows that the constraint function is canonical then it can
+implement a specialized method that directly return the function without calling
+[`Utilities.canonical`](@ref). Therefore, the value returned **cannot** be
+assumed to be a copy of the function stored in `model`.
+Moreover, [`Utilities.Model`](@ref) checks with [`Utilities.is_canonical`](@ref)
+whether the function stored internally is already canonical and if it's the case,
+then it returns the function stored internally instead of a copy.
+"""
+struct CanonicalConstraintFunction <: AbstractConstraintAttribute end
+
+function get_fallback(model::ModelLike, ::CanonicalConstraintFunction, ci::ConstraintIndex)
+    return Utilities.canonical(get(model, ConstraintFunction(), ci))
```

Comment:
Shouldn't the `is_canonical` check go in this fallback? It would save the checks in `Model` and `UniversalFallback`.

Reply:
Done :)